libpromises/evalfunction.c

```diff
@@ -1668,6 +1673,193 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
 
 /*********************************************************************/
 
+#ifdef HAVE_LIBCURL
+static size_t cfengine_curl_write_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+    unsigned int old = BufferSize((Buffer*) userdata);
+    BufferAppend((Buffer*) userdata, ptr, size*nmemb);
+    return BufferSize((Buffer*) userdata) - old;
+}
+#endif
+
+static FnCallResult FnCallCurl(ARG_UNUSED EvalContext *ctx,
+                               ARG_UNUSED const Policy *policy,
+                               ARG_UNUSED const FnCall *fp,
+                               const Rlist *finalargs)
+{
+#ifdef HAVE_LIBCURL
+    if (NULL == CURL_CACHE)
+    {
+        CURL_CACHE = JsonObjectCreate(10);
+    }
+
+    char *url = RlistScalarValue(finalargs);
+    VarRef *ref = ResolveAndQualifyVarName(fp, RlistScalarValue(finalargs->next));
+    if (!ref)
+    {
+        return FnFailure();
+    }
+
+    JsonElement *options = VarRefValueToJson(ctx, fp, ref, NULL, 0);
+    VarRefDestroy(ref);
+
+    if (JsonGetElementType(options) != JSON_ELEMENT_TYPE_CONTAINER ||
+        JsonGetContainerType(options) != JSON_CONTAINER_TYPE_OBJECT)
+    {
+        return FnFailure();
+    }
+
+    Writer *cache_w = StringWriter();
+    WriterWriteF(cache_w, "url = %s; options = ", url);
+    JsonWriteCompact(cache_w, options);
+    JsonElement *old_result = JsonObjectGetAsObject(CURL_CACHE, StringWriterData(cache_w));
+
+    if (NULL != old_result)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s: found cached request for %s", fp->name, url);
+        WriterClose(cache_w);
+        return (FnCallResult) { FNCALL_SUCCESS, (Rval) { JsonCopy(old_result), RVAL_TYPE_CONTAINER } };
+    }
+
+    if (!CURL_INITIALIZED && curl_global_init(CURL_GLOBAL_DEFAULT) != 0)
+    {
+        Log(LOG_LEVEL_ERR, "%s: libcurl initialization failed, sorry", fp->name);
+
+        WriterClose(cache_w);
+        return FnFailure();
+    }
+
+    //TODO: register curl_global_cleanup() to run on agent exit
```

Comment:
Ideally, you'd wrap curl_global_cleanup() in a local function that also frees the cache; use that as tidy-up.
However ...

Your curl activity is currently:
* initialise cache if NULL
* (check argument sanity)
* check for cached answer, use if present
* initialise curl if not already initialised
* set initialised flag [and register tidy-up]

This means that, if curl_global_init fails, you'll still create the cache, that you'll never use [and you'll leak it, even if your tidy-up function would have cleared it away].
That could be avoided by changing the order to:
* (check argument sanity)
* if curl not initialised { initialise it; on success set CURL_INITIALIZED, register tidy-up }
* if no cach, initialise it
* check for cached value, use it.

I don't see any loss of efficiency here, and it'd ensure the tidy-up can also release the cache.
Not urgent, though, as you don't yet have the tidy-up in place ;-)